### ISSUE TITLE
RLN Design Patterns - Practical Development and UX Best Practices

### ISSUE THREAD

**Issue by fryorcraken** [waku-org/pm#102]
_All or part of this issue is likely to become its own milestone and encompass the work tracked with https://github.com/waku-org/pm/issues/71 and https://github.com/waku-org/pm/issues/87_

# Introduction

The usage of RLN for the Waku Network introduces a number of UX and Dev Ex challenges.
Some of the challenges are fundamental in nature (https://github.com/waku-org/research/issues/45), impacting anyone building in a given context. While other are circumstantial (https://github.com/waku-org/pm/issues/71), based on the specific needs of a given application.

This issue attempt to identify a roadmap for the latter.

# Process

1. Identify a use case, clarify the user stories and requirements,
2. Understand user and messages flows and potential issues,
3. Design one or several solution to fulfil criteria, highlight pros and cons of each solution,
4. Implement PoC that uses such solution, confirm assumptions,
5. Identify assets needed for smooth dev ex (SDKs, etc)
6. Produce assets
7. Document usage

For example, in the case of Status Communities (just an example, this are not the actual agreed requirements).

(1) Use case and user stories

Status Communities:
- Owner acts as admin and moderator of a community
- Owner accepts new members via joining request
- Owner wants to pay for member's RLN membership to minimize onboarding friction
- Owner can ban member, and if paid for it, remove member's membership from RLN

(2) user and message flows

```mermaid
sequenceDiagram
	Member->>Owner: request to join
	Owner-->>Member: accept request
```

- How can member contact community owner if they do not have a RLN membership yet?
- How can owner add member to RLN contract?
- How can owner remove member from RLN contract upon ban?
- How can interactions with smart contract remain cheap?
- What privacy guarantees be provided to member? owner?

(3) Solutions

(a) Deploy a Waku node that attach rln proof for incoming light push messages
This node validate that messages are valid community requests before attaching proof
This node runs own traditional DDOS protection (IP rate limit) to not burn through rate limits

cons:
- who run such node? would that expose owner?
considerations:
- if node is censored, member can always pay for own RLN membership

(b) Modify RLN smart contract so that Ethereum address that adds a membership can also remove membership without knowing secret.

cons:
- compromising Eth key means possible RLN membership removal

etc.

(4) Select one or several solution and build a PoC
(5) Learn from PoC, identify SDKs or protocol changes needed to provide turnkey solution. E.g. enable validation plug-in so that such node can easily validate request messages.
(6) Build said SDK/protocols
(7) Document, release

# Use Cases

## A. User Pays for Membership - No Infra model

Most straightforward model where the user pays for their own membership. Developer only write code and ship it, no or minimal hosting from developer (e.g maybe a webpage hosted).
Similar to how one would use Ethereum (pay own gas).

In this model, we do need to review common wallet practices and whether any catering special catering is needed.
- https://github.com/waku-org/research/issues/45 needs to be sorted
- Review wallet models
	- MetaMask extension
	- Status Mobile and Desktop dApp browser
	- Rainbow SDK usage
	- Wallet Connect usage
	- etc

## B. Developers/project pay for Membership - Project Sponsored Model

In this model, the project or developer or ecosystem is funding the RLN membership acquisition.
This could be a model similar to:
- Warpcast: warpcast pays for on-chain (Polygon) actions
- Lens: Tokenomics pay for on-chain (Polygon) actions (need to re-investigate this).

To consider:
- Project may have affinity for a specific L2, do we want to recommend node operator to support this L2 or redirect to a specific L2?
- How does user give their RLN credential to the developers? centralized service? What are the privacy constraints. What can be done to mitigate (usage of push tor, privacy conscious users pay for their own membership, etc).
- How does a user prove their a dApp user? is the model proving ownership of a specific token/NFT or zknote (railgun), how does the project do not end giving RLN memberships to farmers?
- Is staking enabling? is project sponsoring staking or staking+fees?
- Is there a decentralized, smart contract way of doing so? Users pays fees but staking automatically comes out from a project smart contract.
- Is refund/removal needed? If user is not a user anymore (no more zk notes) then should app be able to remove user?

## C> Specific Actors pay for Membership - User Sponsor User model

Basically Status Community model.

## D. User Shares RLN Membership Across Devices or Application

Assuming user has mobile wallet, used wallet to buy RLN membership.
User visit webapp using Waku on Desktop, user wants to use existing RLN membership in webapp.

- How does user sends RLN credentials to webapp?
- What are the security implications? e.g., malicious webapp?

Similar scenarios exists:
- RLN bought in one webapp, to be used in a different webapp/fqdn
- Wallet is desktop application instead of mobile, Waku app is different desktop app or mobile app, etc.

**Comment by fryorcraken** [comment]
An idea regarding sponsorship of RLN membership. we could have a special content topic or pubsub topic where RLN is not applied but instead another rate limit rules are applied.

Message would contain:
- Blockchain address + signature
- blockchain identifier (e.g. chain id)
- RLN ID
- app name or equivalent (first bytes of hash of app names), could be part of content topic

All nodes
- verify that address and signature match
- verify that address has blockchain activity on said chain
- put some sanity rules on app name or equivalent (size check)
- limit number of messages for a given address per day

Sponsoring node (run by developers or special user):
- do same check than other nodes
- confirm user is legit (user has specific on chain activity such as ERC20, NFT, zk note, etc) or maybe whitelisted

Cons:
- blockchain address is linked to RLN ID: this ALMOST is the same than interacting with smart contract. difference is that user has to reveal specific activity. This could be mitigated because user could already be known to use app (e.g sent fund to railgun)
- blockchain address is linked to app: similar issue than above. using hash of app name is a way to introduce k-anonymity.

**Comment by lucille-bellepleure** [comment]
Labs is working on E <3

**Comment by fryorcraken** [comment]
For `C.`, I wonder what would be the security implication of a community owner pre-registering 100 memberships (generated 100 keys locally) and then distribute secrets to community members.
The owner will not use the membership to send messages, as the whole point is to enable their members to send messages.

What information can the owner learn from knowing the secret? @rymnc, thoughts?

**Comment by rymnc** [comment]
If the owner device is compromised, then all participants will get slashed.
I would greatly advice using a stealth commitment scheme that allows a server to generate commitments that only one user may derive private keys for

poc: https://github.com/rymnc/erc-5564-bn254

**Comment by fryorcraken** [comment]
Need to add another pattern in terms of credentials sharing. The simplest use cases/way to frame it would be:

1. How should a user handle several apps? Use the same credentials across apps or get new credentials for each app.
2. How should a user handle several devices? Ditto.

Some concrete use case:
- User gets credentials via mobile phone wallet, can he re-use same credentials in desktop browser web app?
- User gets credentials via one web app, can he use them in another web app? If so, how? What are the security implications? (browser storage is not shared across domains)

**Comment by fryorcraken** [comment]
> (Related to C) For group chat, community owner pays for the RLN

Moving the convo to https://github.com/waku-org/pm/issues/71#issuecomment-1873631088 to not bloat this issue.

**Comment by fryorcraken** [comment]
For `E.` and generally, can a referral system be in place with kick back for users?

**Comment by fryorcraken** [comment]
Idea for `B.` https://github.com/waku-org/pm/issues/148
Where instead of inserting memberships for users, the project attach proofs for users. Particularly relevant when using Waku to ask for wallet signature (tx, auth).

**Comment by igor-sirotin** [comment]
https://github.com/logos-co/anoncomms-pm/issues/17
